### PR TITLE
GH-4282: resolve pulse-wrapper worker count review feedback

### LINKED ISSUE #4282: quality-debt: .agents/scripts/pulse-wrapper.sh — PR #4234 review feedback (medium)
## Unactioned Review Feedback

**Source PR**: #4234
**File**: `.agents/scripts/pulse-wrapper.sh`
**Reviewers**: coderabbit, gemini
**Findings**: 2
**Max severity**: medium

---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/pulse-wrapper.sh:1685`
medium

This chain of `grep` commands can be inefficient, especially in a script that might be called frequently as part of the pulse mechanism. Using a single `awk` command to perform the filtering and counting would be more performant as it avoids creating multiple processes.

```suggestion
	count=$(ps axo command | awk '/\.opencode run/ && /\/full-loop/ && !/\/pulse/ && !/Supervisor Pulse/ {c++} END{print c+0}') || count=0
```

<details>
<summary>References</summary>

1. Optimize shell script pipelines by consolidating multiple filtering commands (like `grep` chains) into a single, more efficient tool like `awk` to improve performance, similar to replacing `grep | sed` with a single `sed` command.
</details>

[View comment](https://github.com/marcusquinn/aidevops/pull/4234#discussion_r2926895535)
---

### MEDIUM: coderabbit (coderabbitai[bot])
**File**: `.agents/scripts/pulse-wrapper.sh:1685`
_⚠️ Potential issue_ | _🟠 Major_

**Make the supervisor exclusion more specific.**

Line 1685 filters out any argv containing `/pulse` or `Supervisor Pulse`, but worker commands include free-form issue/PR text in their title/prompt. That means a real `/full-loop` worker on a pulse-related task can still be excluded here, so utilization remains undercounted for those jobs. Match the supervisor on its dedicated prompt/title shape instead of broad substring drops across the whole command line. As per coding guidelines, automation scripts should focus on reliability and robustness.

[View comment](https://github.com/marcusquinn/aidevops/pull/4234#discussion_r2926907116)
---

---
_Auto-generated by `quality-feedback-helper.sh scan-merged`. Review each finding and either fix the code or dismiss with a reason._

### PR DESCRIPTION
**Pull request by marcusquinn**
## Summary
- Replace the multi-`grep` active worker count pipeline with a single `awk` pass for lower process overhead.
- Narrow supervisor exclusion to the dedicated `--role pulse` + `--session-key supervisor-pulse` signature so `/full-loop` workers mentioning `/pulse` are still counted.
- Add a focused regression test script covering worker counting, supervisor exclusion, and zero-worker behavior.

## Verification
- `shellcheck .agents/scripts/pulse-wrapper.sh`
- `shellcheck .agents/scripts/tests/test-pulse-wrapper-worker-count.sh`
- `bash .agents/scripts/tests/test-pulse-wrapper-worker-count.sh`

Closes #4282

## Summary by CodeRabbit

* **Refactor**
  * Enhanced worker detection logic with more precise filtering to accurately identify active processes.

* **Tests**
  * Introduced comprehensive test coverage for worker counting, including edge cases and non-supervisor configurations.

### CODE REVIEW

**Comment by alex-solovyev** [comment]
Dispatching worker to address unresolved CodeRabbit suggestion.
- **Model**: sonnet (anthropic/claude-sonnet-4-6)
- **Branch**: bugfix/4282-pulse-wrapper-quality-debt
- **Scope**: Address CodeRabbit suggestion: harden supervisor exclusion to argument boundaries (token-boundary matching for --role pulse and --session-key supervisor-pulse)
- **Attempt**: 1 of 1

**Comment by alex-solovyev** [comment]
Addressed CodeRabbit feedback: replaced substring `index()` checks for `--role pulse` and `--session-key supervisor-pulse` with regex token-boundary matching (`/(^|[[:space:]])--role([=[:space:]])pulse([[:space:]]|$)/`) to prevent accidental exclusion of workers whose command text contains these tokens in free-form context.

File: .agents/scripts/tests/test-pulse-wrapper-worker-count.sh [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	fi
+
+	print_result "keeps non-supervisor role pulse commands countable" 1 "Expected 1 active worker, got '${output}'"
+	return 0
+}
+
+main() {
+	setup_test_env
+	trap teardown_test_env EXIT
+
```
[the remark this answers is not in the document]
Comment (alex-solovyev) [reply .agents/scripts/tests/test-pulse-wrapper-worker-count.sh:112]: ✅ Already addressed in commit 26b2b5e — `trap teardown_test_env EXIT` is now registered before `setup_test_env` is called in `main()`.

**Comment by alex-solovyev** [comment]
## Review feedback addressed

Both CodeRabbit suggestions have been applied:

**1. Harden supervisor exclusion to argument boundaries** (nitpick on `pulse-wrapper.sh` lines 1920–1923)
Addressed in commit f4f1088. The `index()` substring checks for `--role pulse` and `--session-key supervisor-pulse` have been replaced with token-boundary regex matches:
```awk
$0 ~ /(^|[[:space:]])--role([=[:space:]])pulse([[:space:]]|$)/ &&
$0 ~ /(^|[[:space:]])--session-key([=[:space:]])supervisor-pulse([[:space:]]|$)/
```
This prevents accidental exclusion of workers whose command text contains these tokens in free-form context (e.g. a worker prompt that mentions `--role pulse` as text).

**2. Register cleanup trap before setup** (actionable comment on `test-pulse-wrapper-worker-count.sh` line 109–112)
Addressed in commit 26b2b5e. `trap teardown_test_env EXIT` is now registered before `setup_test_env` is called, ensuring cleanup runs even if setup fails under `set -e`.
